pepper-apis/dsm-core/src/test/java/org/broadinstitute/dsm/service/admin/UserAdminTestUtil.java

```diff
@@ -0,0 +1,404 @@
+import java.sql.Statement;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import lombok.extern.slf4j.Slf4j;
+import org.broadinstitute.dsm.db.dao.user.UserDao;
+import org.broadinstitute.dsm.db.dto.user.UserDto;
+import org.broadinstitute.dsm.exception.DsmInternalError;
+import org.broadinstitute.lddp.db.SimpleResult;
+import org.junit.Assert;
+
+/**
+ * Provides utility methods to help set up and tear down test realms, test study roles, test users and their roles.
+ * This class keeps track of the entities it creates and will tear them down as requested and/or all at once
+ * when the close() method is called.
+ * If you create users and roles directly with the UserAdminService, you need to manage their lifecycle.
+ * <p></p>
+ * The basic usage is:
+ *  Setup infrastructure (typically in a @BeforeClass or @Before method), but only once per class instance
+ *      Create an instance of this class and call createRealmAndStudyGroup()
+ *      Call setStudyAdminAndRoles() to set up a study user admin and study roles (these include all user roles you plan to assign)
+ *      Note: This class expects actual DSM roles.
+ * <p></p>
+ *  Create users and roles (typically in test)
+ *      Via UserAdminService if you manage the user and role lifecycle
+ *      Or call createTestUser() and methods to add and remove user roles, which are managed by this class
+ * <p></p>
+ *  Delete users and roles (typically in a test or in an @After method)
+ *      Via UserAdminService if you manage the user and role lifecycle
+ *      Or all deleteTestUser() for users managed by this class
+ * <p></p>
+ *  Delete all entities created by this class (typically in an @After or @AfterClass method)
+ *      Call close()
+ */
+@Slf4j
+public class UserAdminTestUtil {
+    private final Map<Integer, List<Integer>> createdUserRoles = new HashMap<>();
+    private final List<Integer> createdGroupRoles = new ArrayList<>();
+    private int studyGroupId = -1;
+    private int ddpInstanceId = -1;
+    private int userAdminRoleId;
+    private int userAdminId = -1;
+    private Map<String, Integer> allRoles;
+
+    private boolean initialized = false;
+
+    private static final String SQL_INSERT_DDP_INSTANCE =
+            "INSERT INTO ddp_instance SET instance_name = ?, is_active = 1, auth0_token = 1, migrated_ddp = 0";
+
+    private static final String SQL_INSERT_DDP_INSTANCE_GROUP =
+            "INSERT INTO ddp_instance_group SET ddp_instance_id = ?, ddp_group_id = ?";
+
+    private static final String SQL_DELETE_DDP_INSTANCE_GROUP =
+            "DELETE FROM ddp_instance_group WHERE ddp_instance_id = ?";
+
+    private static final String SQL_DELETE_DDP_INSTANCE =
+            "DELETE FROM ddp_instance WHERE ddp_instance_id = ?";
+
+
+    public UserAdminTestUtil() {}
+
+    /**
+     * Call this to teardown this class, typically in an @After or @AfterClass method
+     */
+    public void close() {
```

Comment:
How about `deleteGeneratedData()` or a different name that conveys what the method is doing?  I worry that callers won't expect from the name or javadoc that this is erasing test data.

Reply:
It's awesome that this class keep tracks of the data and will wipe it out for you--that will make teardowns super simple.

Reply:
Will do. FWIW, I was attempting to follow the Java close() and closeable() pattern since for Java programmers that's typically how a class releases resources when it goes out of scope. It did not make sense for this class to be autoCloseable() based on how I expect it to be used.